File: output_Codeium/python/Medium/3557-find-maximum-number-of-non-intersecting-substrings.py
# You are given a string word.
# Return the maximum number of non-intersecting substrings of word that are at least four characters long and start and end with the same letter.
# Example 1:
# Input: word = "abcdeafdef"
# Output: 2
# Explanation:
# The two substrings are "abcdea" and "fdef".
# Example 2:
# Input: word = "bcdaaaab"
# Output: 1
# Explanation:
# The only substring is "aaaa". Note that we cannot also choose "bcdaaaab" since it intersects with the other substring.
# Constraints:
# 1 <= word.length <= 2 * 105
# word consists only of lowercase English letters.

class Solution(object):
    def maxSubstrings(self, word):
        """
        :type word: str
        :rtype: int
        """
        count = 0
        n = len(word)
        start = 0

        while start < n:
            end = start
            
            # Find a valid substring starting and ending with the same letter
            while end < n and word[start] != word[end]:
                end += 1
            
            # Check if the substring is long enough
            if end - start + 1 >= 4:
                count += 1
                start = end + 1  # Move to the next potential start
            else:
                start += 1

        return count

File: output_Copilot/java/Medium/3557-find-maximum-number-of-non-intersecting-substrings.java
/*
 You are given a string word.
 Return the maximum number of non-intersecting substrings of word that are at least four characters long and start and end with the same letter.
 Example 1:
 Input: word = "abcdeafdef"
 Output: 2
 Explanation:
 The two substrings are "abcdea" and "fdef".
 Example 2:
 Input: word = "bcdaaaab"
 Output: 1
 Explanation:
 The only substring is "aaaa". Note that we cannot also choose "bcdaaaab" since it intersects with the other substring.
 Constraints:
 1 <= word.length <= 2 * 105
 word consists only of lowercase English letters.
*/

class Solution {
    public int maxSubstrings(String word) {
        int n = word.length();
        int[] last = new int[26];
        for (int i = 0; i < 26; i++) {
            last[i] = -1;
        }
        boolean[] used = new boolean[n];
        int count = 0;

        for (int i = 0; i < n; i++) {
            char c = word.charAt(i);
            int index = c - 'a';
            if (last[index] != -1 && i - last[index] >= 3) {
                count++;
                for (int j = last[index]; j <= i; j++) {
                    used[j] = true;
                }
            }
            last[index] = i;
        }

        for (int i = 0; i < n; i++) {
            if (!used[i]) {
                continue;
            }
            while (i < n && used[i]) {
                i++;
            }
        }

        return count;        
    }
}